pkg/networkservice/chains/client/options.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2021 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"net/url"
+	"time"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/null"
+)
+
+type clientOptions struct {
+	name                    string
+	clientURL               *url.URL
+	cc                      grpc.ClientConnInterface
+	additionalFunctionality []networkservice.NetworkServiceClient
+	authorizeClient         networkservice.NetworkServiceClient
+	refreshClient           networkservice.NetworkServiceClient
+	dialOptions             []grpc.DialOption
+	dialTimeout             time.Duration
+}
+
+// Option modifies default client chain values.
+type Option func(c *clientOptions)
+
+// WithName sets name for the client.
+func WithName(name string) Option {
+	return Option(func(c *clientOptions) {
+		c.name = name
+	})
+}
+
+// WithClientURL sets name for the client.
+func WithClientURL(clientURL *url.URL) Option {
+	return Option(func(c *clientOptions) {
+		c.clientURL = clientURL
+	})
+}
+
+// WithClientConn sets name for the client.
+func WithClientConn(cc grpc.ClientConnInterface) Option {
```

Comment:
I don't quite follow what should be the cases for using this instead of `WithClientURL`. Can you please tell something more about it?

Reply:
Totally... I originally thought as you did, and left this out.   Then I went to test with cmd-forwarder-vpp and discovered that I was using grpc.ClientConnInterface there... and as I tried to unravel that... it turned out to be small and simple to solve here as an option.  It also forced me to allow dial to be flexible for other approaches folks might take for how things get wired up in their chains... basically... dial will dial only if it has a url *and* a grpc.ClientConnInterface has yet to be provided.